### ISSUE TITLE
Make fingerprinting part of FileSource

### ISSUE THREAD

**Issue by Acconut** [tus/tus-js-client#738]
Currently, fingerprints are calculated through the function that is passed in through the `fingerprint` option, defaulting to following implementation:

https://github.com/tus/tus-js-client/blob/ca63ba254ea8766438b9d422f6f94284911f1fa5/lib/browser/fileSignature.js#L12-L27

As can be seen in the code, the fingerprint function has to differentiate between the different types of sources that tus-js-client can handle. For example, it needs to handle React Native files differently than `File`/`Blob` instances. In addition, it does not contain proper handling for streaming sources.

Generating a fingerprint should not be handled in one central function but be delegated to the [FileSource instances](https://github.com/tus/tus-js-client/blob/main/docs/api.md#filereader), which handle the specific access to different input types. These instance know best how to compute the most accurate fingerprint.

That being said, we should still provider users with the ability to influence the fingerprint and add just information in it.

**Comment by major-winter** [comment]
Hi @Acconut , is this issue still relevant? If so, I would like to work on it.

**Comment by Acconut** [comment]
Yes, I would still like to experiment with this idea and see if it provides the anticipated benefits. It's great that you want to help with this! The way I'm currently imagine it is that the `FileSource` interface is extended with a `fingerprint(): Promise<null 
| string>` method. It either return the fingerprint as a string or `null` if no fingerprint can be calculated, e.g. for streams. The `Upload` can then use this method to retrieve a fingerprint. We should still allow users of tus-js-client to customize the fingerprint from FileSource, e.g. my appending just data to it. For example, if uploads shouldn't be reused across projects, users could append a project ID to make sure that fingerprints of the same file for different projects don't match.

Regarding the calculating of the fingerprint, I've also been thinking about making them more precise. Currently, fingerprints only consist about metadata from the file, which can be inaccurate. We should consider including a checksum of the first few megabytes to ensure that the fingerprint reflects the content properly. But we can tackle this in a separate PR as well and keep the current fingerprint logic while moving it to FileSource.

What do you think? Does this make sense?

**Comment by Acconut** [comment]
Sounds good! Let me know if you need any additional help!

**Comment by major-winter** [comment]
Thanks! Also, how can I assign this item to myself? Just a heads-up—I’ll be on a business trip next week, so there might be a slight delay in delivering the feature

**Comment by Acconut** [comment]
Sure, I can assign you for this. And thanks for the heads-up 👍
